content/blog/2017/2017-01-31-pipeline-scalability-best-practice.adoc

```diff
@@ -0,0 +1,197 @@
+--------------------------------
+
+. *Think of pipeline code as glue:* just enough Groovy code to connect
+together the pipeline steps and integrate tools, and no more.
+..   This makes code easier to maintain, more robust against bugs, and
+reduces load on masters.
+.  *Keep it simple:* limit the amount of complex logic embedded in the
+pipeline itself (similarly to a shell script) and avoid treating it as a
+general-purpose programming language.
+..    Pipeline restricts all variables to Serializable types, so keeping
+pipeline logic simple helps avoid a NotSerializableException - see
+appendix at the bottom.
+.  *USE @NonCPS-annotated functions for slightly more complex work.*
+This means more involved processing, logic, and transformations. This
+lets you leverage additional Groovy & functional features for more
+powerful, concise, and performant code.
+..    This still runs on masters so be mindful of complexity, but is much
+faster than native pipeline code because it doesn’t provide durability
+and uses a faster execution model. Still, be mindful of the CPU cost and
+offload to executors for complex work (see below).
+..  NonCPS functions can use a much broader subset of the Groovy
+language, such as iterators and functional features (which makes them
+more terse and fast to write).
+    ..  @NonCPS functions should not use pipeline steps internally, however
+you can store the result of a pipeline step to a variable and use it
+that as the input to a NonCPS function.
+    ... *Gotcha:* It’s not guaranteed that use of a step will generate an
+error (there is an open JIRA to implement that), but you should not rely
+on that behavior. You may see improper handling of exceptions, in
+particular.
+    ..  While normal pipeline is restricted to serializable local variables
+(see appendix at bottom), @NonCPS functions can use more complex,
+nonserializable types internally (ex regex matchers, etc). Parameters
+and return types should still be Serializable, however.
+    ... *Gotcha:* improper usages are not guaranteed to raise an error with
+normal pipeline (optimizations may mask the issue), but it is unsafe to
+rely on this behavior.
+.  *Prefer external scripts/tools for complex or CPU-expensive
+processing* rather than Groovy language features. This offloads work
+from the master to external executors, allowing for easy scale-out of
+hardware resources. It is also generally easier to test because these
+components can be tested in isolation without the full on-master
+execution environment.
+    ..  Many software vendors will provide easy command-line clients for
+their tools in various programming languages. These are often robust,
+performant, and easy to use. Plugins offer another option (see below).
+    ..  Shell or batch steps are often the easiest way to integrate these
+tools, which can be written in any language. For example: sh “java -jar
+client.jar $endPointUrl $inputData” for a Java client, or sh “python
```

Comment:
eh, it's not complex enough for the syntax highlighting to be that important, but I'll set it up as code-block styled